File: src/main/java/com/jackharrhy/storefront/SignListener.java
package com.jackharrhy.storefront;

import org.bukkit.ChatColor;
import org.bukkit.block.*;
import org.bukkit.block.data.BlockData;
import org.bukkit.block.data.Directional;
import org.bukkit.block.data.type.WallSign;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.SignChangeEvent;
import org.bukkit.event.player.PlayerInteractEvent;

public class SignListener implements Listener {
    private Storefront plugin;
    private Storage storage;

    public SignListener(Storefront plugin, Storage storage) {
        this.storage = storage;
        this.plugin = plugin;
        plugin.getServer().getPluginManager().registerEvents(this, plugin);
    }

    private boolean isWallSign(Sign sign) {
        BlockData blockData = sign.getBlockData();
        return blockData instanceof WallSign;
    }

    private Block getBlockBehindWallSign(Sign sign) {
        BlockFace signInverseFace = ((Directional) sign.getBlockData()).getFacing().getOppositeFace();
        Block blockBehindSign = sign.getBlock().getRelative(signInverseFace);
        return blockBehindSign;
    }

    private Chest getChestFromSign(Sign sign) {
        if (!isWallSign(sign)) {
            return null;
        }

        BlockState blockBehindSignState = getBlockBehindWallSign(sign).getState();
        if (!(blockBehindSignState instanceof Chest)) {
            return null;
        }

        return (Chest) blockBehindSignState;
    }

    private void handleNewStorefront(SignChangeEvent event) {
        Block block = event.getBlock();
        BlockState blockState = block.getState();

        if (!(blockState instanceof Sign)) {
            return;
        }

        Sign sign = (Sign) blockState;
        Chest chest = getChestFromSign(sign);

        if (chest != null) {
            this.plugin.updateStorefront(event.getPlayer(), chest, sign);
        }
    }

    @EventHandler
    public void onSignChange(SignChangeEvent event) {
        if (event.getLine(0).toLowerCase().equals("[storefront]")) {
            handleNewStorefront(event);
        }
    }

    @EventHandler
    public void onSignBreak(BlockBreakEvent event) {
        Player player = event.getPlayer();

        Sign sign = getStorefrontSign(event.getBlock());
        if (sign == null) {
            return;
        }

        Chest chest = getChestFromSign(sign);
        if (chest == null) {
            return;
        }

        if (storage.owner(chest.getLocation()).equals(player.getUniqueId().toString())) {
            this.plugin.removeStorefront(player, chest.getLocation());
        } else {
            player.sendMessage(ChatColor.RED + "This isn't your storefront!");
            event.setCancelled(true);
        }
    }

    private Sign getStorefrontSign(Block block) {
        BlockState blockState = block.getState();
        if (!(blockState instanceof Sign)) {
            return null;
        }

        Sign sign = (Sign) blockState;
        if(!sign.getLine(0).equalsIgnoreCase("[storefront]")) {
            return null;
        }

        return sign;
    }

    @EventHandler
    public void onSignInteract(PlayerInteractEvent event){
        Player player = event.getPlayer();

        if (event.getAction() != Action.RIGHT_CLICK_BLOCK) {
            return;
        }

        Sign sign = getStorefrontSign(event.getClickedBlock());
        if (sign == null) {
            return;
        }

        Chest chest = getChestFromSign(sign);

        if (chest != null) {
            this.plugin.updateStorefront(player, chest, sign);
        }
    }
}

File: src/main/kotlin/com/jackharrhy/storefront/SignListener.kt
package com.jackharrhy.storefront

import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer
import org.bukkit.block.Block
import org.bukkit.block.Chest
import org.bukkit.block.Sign
import org.bukkit.block.data.type.WallSign
import org.bukkit.block.sign.Side
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.SignChangeEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.EquipmentSlot

class SignListener(private val plugin: Storefront, private val storage: Storage) : Listener {
    init {
        plugin.server.pluginManager.registerEvents(this, plugin)
    }

    private fun getChestFromSign(sign: Sign): Chest? {
        val wallSign = sign.blockData as? WallSign ?: return null
        return sign.block.getRelative(wallSign.facing.oppositeFace).state as? Chest
    }

    private fun getStorefrontSign(block: Block): Sign? {
        val sign = block.state as? Sign ?: return null
        val firstLine = PlainTextComponentSerializer.plainText().serialize(sign.getSide(Side.FRONT).line(0))
        return sign.takeIf { firstLine.equals("[storefront]", ignoreCase = true) }
    }

    @EventHandler(ignoreCancelled = true)
    fun onSignChange(event: SignChangeEvent) {
        if (event.side != Side.FRONT) return
        val firstLine = event.line(0)?.let { PlainTextComponentSerializer.plainText().serialize(it) }
        if (!firstLine.equals("[storefront]", ignoreCase = true)) return
        val sign = event.block.state as? Sign ?: return
        val chest = getChestFromSign(sign) ?: return
        val playerId = event.player.uniqueId.toString()
        if (storage.ownerUUID(chest.location).orElse(playerId) != playerId) {
            event.isCancelled = true
            return
        }
        // SignChangeEvent fires before the edited text is applied to the block.
        plugin.server.scheduler.runTask(plugin, Runnable {
            if (event.isCancelled) return@Runnable
            val currentSign = getStorefrontSign(event.block) ?: return@Runnable
            val currentChest = getChestFromSign(currentSign) ?: return@Runnable
            plugin.newStorefront(event.player, currentChest, currentSign)
        })
    }

    @EventHandler(ignoreCancelled = true)
    fun onSignBreak(event: BlockBreakEvent) {
        val sign = getStorefrontSign(event.block) ?: return
        val chest = getChestFromSign(sign) ?: return
        event.isCancelled = !plugin.removeStorefront(event.player, chest)
    }

    @EventHandler(ignoreCancelled = true)
    fun onSignInteract(event: PlayerInteractEvent) {
        if (event.action != Action.RIGHT_CLICK_BLOCK || event.hand != EquipmentSlot.HAND) return
        val sign = getStorefrontSign(event.clickedBlock ?: return) ?: return
        val chest = getChestFromSign(sign) ?: return
        plugin.newStorefront(event.player, chest, sign)
    }
}
